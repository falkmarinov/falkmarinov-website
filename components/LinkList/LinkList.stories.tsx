import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';
import { AiOutlineUser } from 'react-icons/ai';

import { LinkList } from './LinkList';

export default {
  title: 'Component/LinkList',
  component: LinkList,
  args: {
    linkItems: [
      {
        icon: <AiOutlineUser />,
        label: 'Example 1',
        url: 'https://example.com/',
      },
      {
        icon: <AiOutlineUser />,
        label: 'Example 2',
        url: 'https://example.com/',
      },
      {
        icon: <AiOutlineUser />,
        label: 'Example 3',
        url: 'https://example.com/',
      },
    ],
  },
} as ComponentMeta<typeof LinkList>;

const Template: ComponentStory<typeof LinkList> = (args) => (
  <LinkList {...args} />
);

export const Default = Template.bind({});
Default.args = {};
