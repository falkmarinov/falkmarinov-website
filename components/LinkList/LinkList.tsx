import React, { useState } from 'react';

import cn from 'classnames';

import Character from '../Character';

interface LinkListItemProps {
  /**
   * Icon of the URL
   */
  icon?: JSX.Element;
  /**
   * Alias which will be shown
   */
  label?: string;
  /**
   * URL of the link
   * TODO: refactor to URL
   */
  url: string;
}

/**
 * Renders a row for LinkList
 */
const LinkListItem = ({ icon, label, url }: LinkListItemProps) => {
  const [isHovered, setIsHovered] = useState(false);

  const onMouseEnterEvent = () => setIsHovered(true);
  const onMouseLeaveEvent = () => setIsHovered(false);

  return (
    <a
      key={label as string}
      className='flex items-center gap-5 group'
      href={url}
      target='_blank'
      rel='noopener noreferrer'
      onMouseEnter={onMouseEnterEvent}
      onMouseLeave={onMouseLeaveEvent}
    >
      <Character isHovered={isHovered}>{icon}</Character>
      <span className={cn(isHovered && 'underline')}>{label}</span>
    </a>
  );
};

interface LinkListProps {
  /**
   *
   */
  linkItems: LinkListItemProps[];
}

export const LinkList = ({ linkItems }: LinkListProps) => {
  return (
    <div className='flex flex-col gap-5'>
      {linkItems.map(({ icon, label, url }, index) => {
        return (
          <LinkListItem
            key={`${index}-${label}`}
            icon={icon}
            label={label}
            url={url}
          />
        );
      })}
    </div>
  );
};
