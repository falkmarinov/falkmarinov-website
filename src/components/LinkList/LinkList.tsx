import React, { useEffect } from 'react';

import cn from 'classnames';
import useHover from 'react-use-hover';
import { AiOutlineLink } from 'react-icons/ai';

import Character from '../Character';

export interface LinkListItemProps {
  /**
   * Icon of the URL
   */
  icon?: JSX.Element;
  /**
   * Alias which will be shown
   */
  label?: string;
  /**
   * URL of the link
   */
  url: string;
  /**
   * Gets called when on enter and on leave
   */
  onHover?: (isHovering: boolean) => void;
}

/**
 * Renders a row for LinkList
 */
const LinkListItem = ({
  icon = <AiOutlineLink />,
  url,
  label = url,
  onHover = () => {},
}: LinkListItemProps) => {
  const [isHovering, hoverProps] = useHover();

  useEffect(() => {
    onHover(isHovering);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isHovering]);

  return (
    <li {...hoverProps}>
      <a
        className='flex items-center gap-5'
        href={url}
        target='_blank'
        rel='noopener noreferrer'
      >
        <Character isHovered={isHovering}>{icon}</Character>
        <div className='relative'>
          <span>{label}</span>
          <span
            className={cn(
              'absolute left-0 transition-all bg-black h-[2px] bottom-1',
              isHovering ? 'w-full' : 'w-0'
            )}
          />
        </div>
      </a>
    </li>
  );
};

interface LinkListProps {
  /**
   * List of links and their icons and labels
   */
  linkItems: LinkListItemProps[];
  /**
   * Gets called when on enter and on leave
   */
  onRowHover?: (index: number | undefined) => void;
}

/**
 * Render a list of links with their icons and labels
 */
export const LinkList = ({
  linkItems,
  onRowHover = () => {},
}: LinkListProps) => {
  return (
    <ol className='flex flex-col gap-5'>
      {linkItems.map(({ icon, label, url }, index) => {
        return (
          <LinkListItem
            key={`${index}-${label}`}
            icon={icon}
            label={label}
            url={url}
            onHover={(isHovering: boolean) => {
              isHovering ? onRowHover(index) : onRowHover(undefined);
            }}
          />
        );
      })}
    </ol>
  );
};
