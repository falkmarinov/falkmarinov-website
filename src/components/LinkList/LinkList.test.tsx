import { render } from '@testing-library/react';

import { FaAddressBook } from 'react-icons/fa';

import { LinkList } from './LinkList';

describe('<LinkList />', () => {
  describe('should render different amounts of items', () => {
    it('One item', () => {
      const { container } = render(
        <LinkList
          linkItems={[
            {
              icon: <FaAddressBook />,
              label: 'Example',
              url: 'https://www.example.com/',
            },
          ]}
        />
      );

      expect(container).toMatchSnapshot();
    });

    it('Two items', () => {
      const { container } = render(
        <LinkList
          linkItems={[
            {
              icon: <FaAddressBook />,
              label: 'Example 1',
              url: 'https://www.example.com/',
            },
            {
              icon: <FaAddressBook />,
              label: 'Example 2',
              url: 'https://www.example.com/',
            },
          ]}
        />
      );

      expect(container).toMatchSnapshot();
    });
  });
});
