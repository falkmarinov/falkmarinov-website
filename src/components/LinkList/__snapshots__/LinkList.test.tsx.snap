// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkList /> should render different amounts of items One item 1`] = `
<div>
  <ol
    class="flex flex-col gap-5"
  >
    <li>
      <a
        class="flex items-center gap-5"
        href="https://www.example.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="flex justify-center items-center w-12 h-12 rounded-full border-2 border-black lg:text-lg transition duration-300 bg-gray-200"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
            />
          </svg>
        </div>
        <div
          class="relative"
        >
          <span>
            Example
          </span>
          <span
            class="absolute left-0 transition-all bg-black h-[2px] bottom-1 w-0"
          />
        </div>
      </a>
    </li>
  </ol>
</div>
`;

exports[`<LinkList /> should render different amounts of items Two items 1`] = `
<div>
  <ol
    class="flex flex-col gap-5"
  >
    <li>
      <a
        class="flex items-center gap-5"
        href="https://www.example.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="flex justify-center items-center w-12 h-12 rounded-full border-2 border-black lg:text-lg transition duration-300 bg-gray-200"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
            />
          </svg>
        </div>
        <div
          class="relative"
        >
          <span>
            Example 1
          </span>
          <span
            class="absolute left-0 transition-all bg-black h-[2px] bottom-1 w-0"
          />
        </div>
      </a>
    </li>
    <li>
      <a
        class="flex items-center gap-5"
        href="https://www.example.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="flex justify-center items-center w-12 h-12 rounded-full border-2 border-black lg:text-lg transition duration-300 bg-gray-200"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
            />
          </svg>
        </div>
        <div
          class="relative"
        >
          <span>
            Example 2
          </span>
          <span
            class="absolute left-0 transition-all bg-black h-[2px] bottom-1 w-0"
          />
        </div>
      </a>
    </li>
  </ol>
</div>
`;
